lib/ruby_lsp/requests/support/rails_document_client.rb

```diff
@@ -0,0 +1,123 @@
+
+module RubyLsp
+  module Requests
+    module Support
+      class RailsDocumentClient
+        RAILS_DOC_HOST = "https://api.rubyonrails.org"
+        SUPPORTED_RAILS_DOC_NAMESPACES = T.let(
+          Regexp.union(
+            /ActionDispatch/, /ActionController/, /AbstractController/, /ActiveRecord/, /ActiveModel/, /ActiveStorage/,
+            /ActionText/, /ActiveJob/
+          ).freeze,
+          Regexp,
+        )
+
+        RAILTIES_VERSION = T.let(
+          [*::Gem::Specification.default_stubs, *::Gem::Specification.stubs].find do |s|
+            s.name == "railties"
+          end&.version&.to_s, T.nilable(String)
+        )
+
+        class << self
+          extend T::Sig
+          sig { returns(T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]])) }
+          def search_index
+            @rails_documents ||= T.let(
+              build_search_index,
+              T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]]),
+            )
+          end
+
+          sig { returns(T.nilable(T::Hash[String, T::Array[T::Hash[Symbol, String]]])) }
+          private def build_search_index
+            return unless RAILTIES_VERSION
+
+            $stderr.puts "Fetching Rails Documents..."
+            # If the version's doc is not found, e.g. Rails main, it'll be redirected
+            # In this case, we just fetch the latest doc
+            response = if Gem::Version.new(RAILTIES_VERSION).prerelease?
+              Net::HTTP.get_response(URI("#{RAILS_DOC_HOST}/js/search_index.js"))
+            else
+              Net::HTTP.get_response(URI("#{RAILS_DOC_HOST}/v#{RAILTIES_VERSION}/js/search_index.js"))
+            end
+
+            if response.code == "200"
+              process_search_index(response.body)
+            else
+              $stderr.puts("Response failed: #{response.inspect}")
+              nil
+            end
+          rescue StandardError => e
+            $stderr.puts("Exception occurred when fetching Rails document index: #{e.inspect}")
+          end
+
+          sig { params(js: String).returns(T::Hash[String, T::Array[T::Hash[Symbol, String]]]) }
+          private def process_search_index(js)
+            table = {}
+            raw_data = js.sub("var search_data = ", "")
+            info = JSON.parse(raw_data).dig("index", "info")
+
+            # An entry looks like this:
+            #
+            # ["belongs_to",                                                              # method or module/class
+            #  "ActiveRecord::Associations::ClassMethods",                                # method owner
+            #  "classes/ActiveRecord/Associations/ClassMethods.html#method-i-belongs_to", # path to the document
+            #  "(name, scope = nil, **options)",                                          # method's parameters
+            #  "<p>Specifies a one-to-one association with another class..."]             # document preview
+            #
+            info.each do |ary|
```

Comment:
I should have written this on the previous PR as well but thank you for using comments to make this code so easy to understand. The documentation is so valuable!